src/notify/notify.c

```diff
@@ -313,56 +338,238 @@ static int parse_argv(int argc, char *argv[]) {
+
+        _cleanup_free_ char *addr_string = NULL;
+        r = getsockname_pretty(socket_fd, &addr_string);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get socket name: %m");
+
+        _cleanup_free_ char *c = strv_join(command, " ");
+        if (!c)
+                return log_oom();
+
+        _cleanup_(pidref_done) PidRef child = PIDREF_NULL;
+        r = pidref_safe_fork_full(
+                        "(notify)",
+                        /* stdio_fds= */ (const int[]) { -EBADF, -EBADF, STDERR_FILENO },
+                        /* execpt_fds= */ NULL,
+                        /* n_except_fds= */ 0,
+                        /* flags= */ FORK_REARRANGE_STDIO,
+                        &child);
+        if (r < 0)
+                return log_error_errno(r, "Failed to fork child in order to execute '%s': %m", c);
+        if (r == 0) {
+                socket_fd = safe_close(socket_fd);
+                pidref_done(&child);
+
+                if (setenv("NOTIFY_SOCKET", addr_string, /* overwrite= */ true) < 0) {
+                        log_error_errno(errno, "Failed to set $NOTIFY_SOCKET: %m");
+                        _exit(EXIT_MEMORY);
+                }
+
+                log_debug("Executing: %s", c);
+                execvp(command[0], command);
+                log_error_errno(errno, "Failed to execute '%s': %m", c);
+                _exit(EXIT_EXEC);
+        }
+
+        if (!arg_quiet) {
+                printf(PID_FMT "\n", child.pid);
+                fflush(stdout);
+        }
+
+        BLOCK_SIGNALS(SIGCHLD);
+
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        r = sd_event_new(&event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate event loop: %m");
+
+        r = sd_event_add_io(event, /* ret_event_source= */ NULL, socket_fd, EPOLLIN, on_notify_socket, &child);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate IO source: %m");
+
+        _cleanup_(sd_event_source_unrefp) sd_event_source *s = NULL;
+        r = event_add_child_pidref(event, &s, &child, WEXITED, on_child, /* userdata= */ NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate child source: %m");
+
+        /* If we receive both the sd_notify() message and a SIGCHLD always process sd_notify() first, it's
+         * the more interesting, "positive" information. */
+        r = sd_event_source_set_priority(s, SD_EVENT_PRIORITY_NORMAL + 10);
+        if (r < 0)
+                return log_error_errno(r, "Failed to change child event source priority: %m");
+
+        sd_event_source **forward_signal_sources = NULL;
+        size_t n_forward_signal_sources = 0;
+        CLEANUP_ARRAY(forward_signal_sources, n_forward_signal_sources, event_source_unref_many);
+
+        const int forward_signals[] = { SIGTERM, SIGINT, SIGQUIT, SIGTSTP, SIGCONT, SIGUSR1, SIGUSR2 };
```

Comment:
static?

Reply:
Shouldn't we forward SIGHUP here as well since the child won't receive it? Or at least handle it gracefully ourselves?